### PR TITLE
Command line interface

### LINKED ISSUE #7: A command line interface
It would be nice if I could call `markdownlint` from my NPM/build scripts... :)

### PR DESCRIPTION
**Pull request by igorshubovych**
- commander package to run commands.
- .editorconfig

I can also add the info about it in README.md

It fixes #7.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,9 @@
+root = true
+
+[*]
+indent_style = space
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
```
Comment (jawshooah) [inline .editorconfig]: Ironic that this file is missing a trailing newline...right after `insert_final_newline = true` :wink:

**Comment by DavidAnson** [comment]
Wow, this looks great!

I've been thinking about how to add command-line support ever since #7 was opened. This is definitely something I'd like to do, but I've been trying to balance it against my goal to have a minimal set of dependencies (currently just markdown-it). I figured that CLI support would require at least a couple of new dependencies and wanted to avoid that for scenarios that don't take advantage of it (i.e., all current uses). What I eventually decided was to create a separate project/package for the CLI that depends on this project.

I had planned to create that project next, but before I got a chance to do so you submitted this pull request. :) I think what you've done is fantastic, and am not sure about the new dependencies, so I have an idea: What do you think about creating a separate project/package for the CLI based on what you've already written? I would link to your CLI from this README and anyone who's interested would be able to install it!

If you like the idea, I already have a couple of (minor) feature requests ready. :) Or if you do not want to take on ownership of a new project, I could create one and you could resubmit your pull request there.

Please let me know what think - I'm excited to have a command-line interface to markdownlint and think we can make either of these approaches successful!

**Comment by jawshooah** [comment]
> wanted to avoid that for scenarios that don't take advantage of it (i.e., all current uses)

Well, no current uses take advantage of it by definition, since it doesn't exist yet :wink:

A CLI would be very useful for non-node projects that don't have a grunt/gulp build workflow. For example, [overcommit](https://github.com/brigade/overcommit) could benefit from a pre-commit hook that runs `markdownlint` on modified Markdown files (though I've already [opened a PR](https://github.com/brigade/overcommit/pull/323) to do the same with [mdl](https://github.com/mivok/markdownlint)).

**Comment by jawshooah** [comment]
[SublimeLinter](https://github.com/SublimeLinter/SublimeLinter3) could also benefit from a `markdownlint` plugin, which could only exist if a CLI were provided.

**Comment by DavidAnson** [comment]
To be clear, I'm totally in support of adding a CLI for markdownlint! My comment above is just that some scenarios don't need the CLI, so I'd rather avoid forcing the corresponding dependencies onto those scenarios. (Note that in this case the number of required dependencies went from 1 to 5.) Having a separate project for the CLI is a nice way to have your cake and eat it, too. :)

**Comment by igorshubovych** [comment]
I probably can rewrite it with argparse instead of commander. argparse is already used by markdown-it. But it also uses lodash.

**Comment by igorshubovych** [comment]
However there are high chances that commander, rc and so lodash and deep-extend as their dependencies are already used in any node project, because they are often used in some linters (similar to markdownlint) and similar projects.

**Comment by DavidAnson** [comment]
I tried to do a bit of research on this topic today; I found just one discussion of the topic (please link to others if you find them): sindresorhus/ama#17

My read of the comments in that issue (as well as the linked issues) is that there good reasons to have a separate project for the CLI. So that remains my preference for markdownlint.

That said, if you have thoughts on why the CLI *should* be bundled together with markdownlint, I'm keeping an open mind. :)

Thanks for the interesting discussion!

**Comment by igorshubovych** [comment]
That's was definitely an interesting discussion. Thanks for sharing.

So I published version 0.0.1 of CLI for it.
https://www.npmjs.com/package/markdownlint-cli
https://github.com/igorshubovych/markdownlint-cli

I promise to continue working on it.

@DavidAnson do you want me to add you as a collaborator?

**Comment by DavidAnson** [comment]
Awesome! :)

I will link to your project from the README later today and give it a try as well.

Regarding the ideas I had, would you like me to create issues for them in your project, mention them here, or something else?

> do you want me to add you as a collaborator?

I would be honored if you did, but it's completely up to you.

**Comment by igorshubovych** [comment]
> Regarding the ideas I had, would you like me to create issues for them in your project, mention them here, or something else?
Whatever you prefer. `markdownlint-cli` is just a thin wrapper over `markdownlint`, so if the issue is related to it, then yes, definitely create an issue.

BTW, I added you as a collaborator on Github repo and npm package. Just in case I disappear, and something needs to be addressed.

**Comment by igorshubovych** [comment]
So I am closing this one.

**Comment by DavidAnson** [comment]
Excellent, thank you!

I created a few issues from the ideas I had - you're welcome to agree or disagree with each of them. :)

Also, thanks for the pull request to update the README, I merged it and moved markdownlint-cli to the top of that section.

**Comment by DavidAnson** [comment]
*By the way, I will tweet markdownlint 0.1.0 tomorrow and then markdownlint-cli the next day.*

**Pull request closed without merging**